functions/vshn-postgres-func/encrypted_pvc.go

```diff
@@ -0,0 +1,72 @@
+package vshnpostgres
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/sethvargo/go-password/password"
+
+	"github.com/go-logr/logr"
+	"github.com/vshn/appcat-comp-functions/runtime"
+	vshnv1 "github.com/vshn/component-appcat/apis/vshn/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AddPvcSecret adds a secret for the encrypted PVC for the  PostgreSQL instance.
+func AddPvcSecret(ctx context.Context, log logr.Logger, iof *runtime.Runtime, comp *vshnv1.VSHNPostgreSQL) (*vshnv1.VSHNPostgreSQL, error) {
+
+	log.Info("Check if encrypted storage is enabled")
+
+	log.V(1).Info("Transforming", "obj", iof)
+
+	// luksSecretResourceName is the resource name defined in the composition
+	// This name is different from metadata.name of the same resource
+	// The value is hardcoded in the composition for each resource and due to crossplane limitation
+	// it cannot be matched to the metadata.name
+	var luksSecretResourceName = comp.Name + "-luks-key"
+
+	encryptionSpec := comp.Spec.Parameters.Encryption
+
+	if _, err := strconv.ParseBool(encryptionSpec.Enabled); err == nil {
+
+		log.Info("Adding secret to composite")
+
+		s := &v1.Secret{}
+
+		err := iof.GetFromObservedKubeObject(ctx, s, luksSecretResourceName)
+		var luksKey string
+		if err != nil {
+			log.Info("Secret does not exist yet. Creating...")
+			luksKey, _ = password.Generate(64, 10, 1, false, true)
+		} else {
+			log.Info("Secret already exists. Use that one")
+			luksKey = string(s.Data["luksKey"])
+		}
+		log.Info("Foo: " + luksKey)
+		secret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-data-%s-0-luks-key", comp.ObjectMeta.Labels["crossplane.io/composite"], comp.ObjectMeta.Labels["crossplane.io/composite"]),
+				Namespace: comp.Status.InstanceNamespace,
+			},
+			Data: map[string][]byte{
+				"luksKey": []byte(luksKey),
+			},
+		}
```

Comment:
Nitpick: This would IMO be a bit cleaner

```suggestion
		secret := &v1.Secret{}

		err := iof.GetFromObservedKubeObject(ctx, s, luksSecretResourceName)
		var luksKey string
		if err != nil {
			log.Info("Secret does not exist yet. Creating...")
			luksKey, _ = password.Generate(64, 10, 1, false, true)
		
		    secret = &v1.Secret{
			    ObjectMeta: metav1.ObjectMeta{
				    Name:      fmt.Sprintf("%s-data-%s-0-luks-key", comp.ObjectMeta.Labels["crossplane.io/composite"], comp.ObjectMeta.Labels["crossplane.io/composite"]),
				    Namespace: comp.Status.InstanceNamespace,
			    },
			    Data: map[string][]byte{
				    "luksKey": []byte(luksKey),
			    },
		    }
		}
```